tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -2558,6 +2561,75 @@ void CGMSHLSLRuntime::AddHLSLFunctionInfo(Function *F, const FunctionDecl *FD) {
   m_ScopeMap[F] = ScopeInfo(F, FD->getLocation());
 }
 
+// Find the input node record field with the SV_DispatchGrid semantic.
+// We have already diagnosed any error conditions in Sema, so we
+// expect valid size and types, and use the first occurance found.
+// We return true if we have populated the SV_DispatchGrid values.
+bool CGMSHLSLRuntime::FindDispatchGridSemantic(const CXXRecordDecl *RD,
+                                               hlsl::SVDispatchGrid &SDGRec,
+                                               CharUnits Offset) {
+  const ASTRecordLayout &Layout = CGM.getContext().getASTRecordLayout(RD);
+
+  // Collect any bases.
+  SmallVector<const CXXRecordDecl *, 4> Bases;
+  for (const CXXBaseSpecifier &Base : RD->bases()) {
+    if (!Base.getType()->isDependentType())
+      Bases.push_back(Base.getType()->getAsCXXRecordDecl());
+  }
+
+  // Sort bases by offset.
+  std::stable_sort(Bases.begin(), Bases.end(),
+                   [&](const CXXRecordDecl *L, const CXXRecordDecl *R) {
+                     return Layout.getBaseClassOffset(L) <
+                            Layout.getBaseClassOffset(R);
+                   });
```

Comment:
What does ordering the bases give us? I could imagine this might be intended to prevent duplicate testing, but If there are multiple levels of bases, it seems this results in checking the highest parent base once for each descendent.

Reply:
We need to order the bases so we can keep track of the offset from the start of the input record.
If a record type appears more than once then it can't contain a field with SV_DispatchGrid as that would be errored in Sema, so we could track which records we have already checked and skip checking the fields of any we have seen before. However, I'm not sure that such cases would occur often enough to make the extra housekeeping worthwhile - so currently we may check records more than once.